edgelet/edgelet-docker/src/migc_persistence.rs

```diff
@@ -0,0 +1,506 @@
+use std::io::Write;
+use std::sync::{Arc, Mutex};
+use std::time::UNIX_EPOCH;
+use std::{collections::HashMap, fs, time::Duration};
+
+use edgelet_settings::base::image::MIGCSettings;
+
+use crate::Error;
+
+#[derive(Debug, Clone)]
+struct MIGCPersistenceInner {
+    filename: String,
+    settings: MIGCSettings,
+}
+
+#[derive(Debug, Clone)]
+pub struct MIGCPersistence {
+    inner: Arc<Mutex<MIGCPersistenceInner>>,
+}
+
+impl MIGCPersistence {
+    pub fn new(filename: String, settings: Option<MIGCSettings>) -> Self {
+        let settings = match settings {
+            Some(settings) => settings,
+            None => MIGCSettings::new(Duration::MAX, Duration::MAX, false),
+        };
+
+        Self {
+            inner: Arc::new(Mutex::new(MIGCPersistenceInner { filename, settings })),
+        }
+    }
+
+    /// # Panics
+    pub async fn record_image_use_timestamp(&self, name_or_id: &str) {
+        let guard = self
+            .inner
+            .lock()
+            .expect("Could not lock images file for image garbage collection");
+
+        let migc_filename = guard.filename.clone();
```

Comment:
It's not _filename_ that's guarded, per se. It's _access_ to 'name of file to be opened' to prevent potential synchronization issues.

If you can't access _filename_ without a mutex, you can't introduce races.

There are only two use-cases for which access to _filename_ is required:

1. To add image to file (entails reading file, and adding/updating image timestamp -- RW access, basically) 
2. Prune images off disk (entails reading file, figuring out what needs deletion, and then writing images that won't be deleted back to file -- RW access again)

So, since access in both use cases is always RW, there's no current use case where multiple readers would require read-only access.


Finally, it's not clear to me how we would guard access to the file, if not by restricting access to _filename_. For instance, I could use this library: https://docs.rs/fs2/0.4.2/fs2/trait.FileExt.html    BUT it comes with the following caveats **in addition to taking a dependency on that library**:

"File locks are a cross-platform hazard since the file lock APIs exposed by operating system kernels vary in subtle and not-so-subtle ways.

The API exposed by this library can be safely used across platforms as long as the following rules are followed:

- Multiple locks should not be created on an individual File instance concurrently.
- Duplicated files should not be locked without great care.
- Files to be locked should be opened with at least read or write permissions.
- File locks may only be relied upon to be advisory."

Unless performance (or correctness) is demonstrably shown to be affected by _filename_ being guarded by the mutex, it's just easier to not have to worry about all this.

Definitely open to other/better ways though.